src/pprof/ddprof_pprof.cc

```diff
@@ -34,54 +32,77 @@ std_string_2_slice_c_char(const std::string &str) {
 #define SLICE_LITERAL(str)                                                     \
   (struct ddprof_ffi_Slice_c_char) { .ptr = (str), .len = sizeof(str) - 1 }
 
+#define PPROF_MAX_LABELS 5
+
 static ddprof_ffi_Slice_c_char ffi_empty_char_slice(void) {
   return (struct ddprof_ffi_Slice_c_char){.ptr = (NULL), .len = 0};
 }
 
-static const struct ddprof_ffi_ValueType s_sample_count = {
-    .type_ = SLICE_LITERAL("sample"),
-    .unit = SLICE_LITERAL("count"),
-};
-
-static const struct ddprof_ffi_ValueType s_cpu_time = {
-    .type_ = SLICE_LITERAL("cpu-time"),
-    .unit = SLICE_LITERAL("nanoseconds"),
-};
-
-static const struct ddprof_ffi_ValueType s_cpu_sample = {
-    .type_ = SLICE_LITERAL("cpu-sample"),
-    .unit = SLICE_LITERAL("count"),
-};
-
-DDRes pprof_create_profile(DDProfPProf *pprof, const PerfOption *options,
-                           unsigned nbOptions) {
-  // Create one value
-  struct ddprof_ffi_ValueType perf_value_type[MAX_TYPE_WATCHER + 1];
-  unsigned value_index = 0;
-  perf_value_type[value_index++] = s_sample_count;
-
-  /* Add one value type per watcher */
-  for (; value_index < nbOptions + 1; ++value_index) {
-    ddprof_ffi_Slice_c_char perf_value_slice = {.ptr = options->label,
-                                                .len = strlen(options->label)};
-    ddprof_ffi_Slice_c_char perf_unit = {.ptr = options->unit,
-                                         .len = strlen(options->unit)};
-
-    perf_value_type[value_index].type_ = perf_value_slice;
-    perf_value_type[value_index].unit = perf_unit;
+DDRes pprof_create_profile(DDProfPProf *pprof, DDProfContext *ctx) {
+  // Figure out which sample_type_ids are used by active watchers
+  const PerfWatcher *watchers = ctx->watchers;
+  size_t num_watchers = ctx->num_watchers;
+  bool active_ids[DDPROF_PWT_LENGTH] = {};
+  struct ddprof_ffi_ValueType perf_value_type[DDPROF_PWT_LENGTH];
+  for (unsigned i = 0; i < num_watchers; ++i) {
+    int this_id = watchers[i].sample_type_id;
+    if (this_id < 0 || this_id == DDPROF_PWT_NOCOUNT || this_id >= DDPROF_PWT_LENGTH) {
+      LG_WRN("Watcher \"%s\" (%d) has invalid sample_type_id %d, ignoring",
+          watchers[i].desc, i, this_id);
+      continue;
+    }
+    active_ids[watchers[i].sample_type_id] = true;
+  }
+
+  // Now that we have all of the sample ids, we populate the sample_type_id 
+  // permutation vector. This allows the exporter to take the absolute sample ID 
+  // and convert it into the index of the pprof.
+  memset(ctx->sample_type_pv, 0, sizeof(*ctx->sample_type_pv)*DDPROF_PWT_LENGTH);
```

Comment:
`sample_type_pv` should probably be initialized to something different from 0, since 0 is a valid value.

Reply:
It's a valid enum value, but it it should be guarded against later.  Somewhat of a flavor question.  One preference is to do absolute boundary checks, the other is to check against a sentinel value.  Still not sure, but in the new code the NOCOUNT profile type is checked against explicitly and the scope of the `pv` is limited to this function alone.